cloudinit/lifecycle.py

```diff
@@ -0,0 +1,237 @@
+    ) -> "Version":
+        """Default of -1 allows us to tiebreak in favor of the most specific
+        number"""
+        return super(Version, cls).__new__(cls, major, minor, patch, rev)
+
+    @classmethod
+    def from_str(cls, version: str) -> "Version":
+        """Create a Version object from a string.
+
+        :param version: A period-delimited version string, max 4 segments.
+
+        :raises TypeError: Raised if invalid arguments are given.
+        :raises ValueError: Raised if invalid arguments are given.
+
+        :return: A Version object.
+        """
+        return cls(*(list(map(int, version.split(".")))))
+
+    def __gt__(self, other):
+        return 1 == self._compare_version(other)
+
+    def __eq__(self, other):
+        return (
+            self.major == other.major
+            and self.minor == other.minor
+            and self.patch == other.patch
+            and self.rev == other.rev
+        )
+
+    def __iter__(self):
+        """Iterate over the version (drop sentinels)"""
+        for n in (self.major, self.minor, self.patch, self.rev):
+            if n != -1:
+                yield str(n)
+            else:
+                break
+
+    def __str__(self):
+        return ".".join(self)
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def _compare_version(self, other: "Version") -> int:
+        """Compare this Version to another.
+
+        :param other: A Version object.
+
+        :return: -1 if self > other, 1 if self < other, else 0
+        """
+        if self == other:
+            return 0
+        if self.major > other.major:
+            return 1
+        if self.minor > other.minor:
+            return 1
+        if self.patch > other.patch:
+            return 1
+        if self.rev > other.rev:
+            return 1
+        return -1
+
+
+def should_log_deprecation(version: str, boundary_version: str) -> bool:
+    """Determine if a deprecation message should be logged.
+
+    :param version: The version in which the thing was deprecated.
+    :param boundary_version: The version at which deprecation level is logged.
+
+    :return: True if the message should be logged, else False.
+    """
+    return boundary_version == "devel" or Version.from_str(
+        version
+    ) <= Version.from_str(boundary_version)
+
+
+def log_if_acceptable(
+    *, logger: logging.Logger, version: str, level: int, msg: str, args
+):
+    """Log a message at the requested level, if that is acceptable.
+
+    If the log level is too high due to the version boundary, log at DEBUG
+    level. Useful to add new warnings to previously unguarded code without
+    disrupting stable downstreams.
+
+    :param logger: Logger object to log with
+    :param version: Version string of the version that this log was introduced
+    :param level: Preferred level at which this message should be logged
+    :param msg: Message, as passed to the logger.
+    :param args: Message parameters
```

Comment:
i feel like this is somewhat better for clarity but maybe others feel it is unnecessary 
```suggestion
    :param args: Message formatting args, as passed to the logger
```